### PR TITLE
Add package `DocBlockr Python`

### PR DESCRIPTION
**Pull request by adambullmer**

### CODE REVIEW

**Comment by FichteFoll** [comment]
https://github.com/KristoforMaynard/SublimeAutoDocstring does this too. It supports different styles, including `'google'`, so I propose looking there first.

Additionally, did you consider adding this to DocBlockr directly? If yes, why didn't you? https://github.com/spadgos/sublime-jsdocs/issues/113

cc @spadgos

**Comment by adambullmer** [comment]
**TL;DR** _Python is so drastically different than C based languages that it would require an almost entirely new plugin, or at the very least a rewrite to the inner workings of jsdocs to make it possible._

I thought about trying to create a new parser to fit in the jsdocss schema, but decided to roll my own to get it up and running first. Python offers a few more features that would require their package to be a little more robust: Keyword arguments, Decorators, Extending Classes, Yielding. It would also require a document parser that will read in both directions of the docstring: above for decorators and module definitions, and below to try and interpret class / module variables, return / yield values. jsdoc determines scope by brackets, as is the convention of the languages it currently supports, but python is an indentation based language, and needs different considerations to determine end of function or class.

I'm not opposed to contributing to jsdocs as it would greatly benefit this project, but I think this plugin needs more time to mature and be more feature complete. Also, a lot of collaborative work from the guys at @spadgos would be needed to rework the basic components they have established to support non C based languages.

I tried AutoDocstring, but I think it isn't quite as feature complete for generating the docstring tempalte as I was hoping this plugin to be, and would definitely prefer the plugin to function more like the jsdocs plugin does.

**Comment by FichteFoll** [comment]
I agree with your reasoning. Maybe @spadgos has something to say about this, too?

>I tried AutoDocstring, but I think it isn't quite as feature complete for generating the docstring tempalte as I was hoping this plugin to be

What prevents you from filing a pull request against AutoDocstring?

>and would definitely prefer the plugin to function more like the jsdocs plugin does.

What do you mean exactly?

**Comment by FichteFoll** [comment]
(submitted too early)

If you can, please try to extend existing packages first before working on your own. There are reasons that would prevent collaborating on a project, but in most cases it's just a combination of laziness or unwillingness to contribute to a non-owned repository. See bottom for copy-pasted reason for preventing duplicates

cc @KristoforMaynard 

---

We try to keep duplicated packages at a minimum because of several reasons, mainly:

1. If the feature provided is trivial, the user is prompted with different packages all achieving practically the same thing and confused about which he should use.
2. If the difference is trivial, the user is again confused about which he should use, but the package authors or maintainers are also effectively doubling their work (of maintainance). If they worked together on the same package (either by collaborating or by proposing pull requests) the community would be.
3. If one pf the authors or maintainers can not keep maintaining his version of a package it is abandoned. With two or more authors collaborating the chance of this happening is reduced.

There are exceptions to this rule in cases where developers simply can't or don't want to work together due to certain reasons or the packages themselves are different enough in that they provide different use cases and would attract different users. However, there might also still be the option of combining these kind of packages by including functionality of one in the other.

However, there is no definitive objective criteria here.

**Comment by spadgos** [comment]
It's true that it would take a significant effort to get non c-style languages supported in docblockr.

*completely tangential side note: you should consider using non-gendered pronouns when speaking of people of unknown gender (eg: users). Instead of 'he'/'his' just use 'they'/'their'.*

**Comment by KristoforMaynard** [comment]
After a (very) cursory glance at the `DocBlockr Python` plugin, the functionality that I saw that it supplies in addition to those already in `AutoDocstring` are:

1. Inserting a snippet instead of editing the buffer. From a usability perspective I like this approach and think that would be a great feature to merge into AutoDocstring.
2. Extra parsing, such as pulling out the parent class from a class defintion. This seems useful.
3. An alternative (not Google or Numpy) formatting for the docstring.

If there is interest, I would love to discuss pull requests into `AutoDocstring` for these features, and ones that I no doubt missed.

**Comment by abathur** [comment]
@spagdos :+1: for the pronouns aside.

**Comment by FichteFoll** [comment]
I already fixed my template, so thank you for the aside, @spadgos.

Regarding this PR, @adambullmer, please get in contact with @KristoforMaynard and work on the existing plugin to make it the best Python docstring generator Sublime Text has ever seen.

Thus, I'm closing this.
